dd-java-agent/instrumentation/jersey/src/main/java/datadog/trace/instrumentation/jersey/JerseyRequestExtractAdapter.java

```diff
@@ -0,0 +1,38 @@
+package datadog.trace.instrumentation.jersey;
+
+import io.opentracing.propagation.TextMap;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import javax.ws.rs.core.MultivaluedMap;
+import org.glassfish.jersey.server.ContainerRequest;
+
+public class JerseyRequestExtractAdapter implements TextMap {
+  private final MultivaluedMap headers;
+
+  JerseyRequestExtractAdapter(final ContainerRequest request) {
+    headers = request.getHeaders();
+  }
+
+  @Override
+  public Iterator<Map.Entry<String, String>> iterator() {
+    HashMap<String, String> delinkedListHeaders = new HashMap<>();
+
+    // Jersey headers are Map<String, List<String>> but B3 only supports single value
+    for (final Object e : headers.entrySet()) {
+      Map.Entry<String, List<String>> entry = (Map.Entry<String, List<String>>) e;
+      final String key = entry.getKey().toLowerCase();
+      // first item will do.
+      final String val = entry.getValue().get(0);
+      delinkedListHeaders.put(key, val);
+    }
+
+    return delinkedListHeaders.entrySet().iterator();
```

Comment:
Another option altogether is to wrap the `headers.entrySet().iterator()` instead of building a new map:

```java
return new Iterator<Map.Entry<String, String>>() {
    private final Iterator<Map.Entry<String, List<String>>> iterator = headers.entrySet().iterator();
    public boolean hasNext() { return iterator.hasNext(); }
    public boolean remove() { return iterator.remove(); }
    public Map.Entry<String, String> next() {
      Map.Entry<String, List<String>> entry = iterator.next();
      return new AbstractMap.Entry<String, String>(entry.getKey(), entry.getValue().get(0));
    }
};
```

Reply:
Updated w/ this approach, but there was some kind of helper class injector conflict when using an anonymous inner class.  I'll need to look into that later.